### PR TITLE
Enable EXPORT_ALL in main modules with the wasm backend

### LINKED ISSUE #9793: MAIN_MODULE without EXPORT_ALL does not export js library functions
See for example the testcase added in https://github.com/emscripten-core/emscripten/pull/9698 - we need EXPORT_ALL there so that the side module can call `cxa_throw`, which is a JS library function.

It shouldn't be an observable difference if a function is from the JS library or not, so it should be exported like the compiled library code is, probably. I'm not sure of the ramifications of flipping EXPORT_ALL to 1 when MAIN_MODULE, though, so maybe we need something more specific?

### PR DESCRIPTION
**Pull request by kripken**
Fastcomp adds all the JS library to the asmLibraryArg object which is
provided as "env" to wasm modules, which lets them access those methods.
In the wasm backend we use a more straightforward method of flipping
the `EXPORT_ALL` flag, which provides those methods on Module, which is
used to look up imports as necessary.

`EXPORT_ALL` is also noticed by metadce (when present, it doesn't try to
remove wasm exports) so this likely fixes another bug as well, see the
metadce output changes (the additions were metadce'd out before,
possibly incorrectly if a side module needs them).

Fixes #9793

### CODE REVIEW

**Comment by sbc100** [comment]
I'm confused though because you explicitly disabled EXPORT_ALL for MAIN_MODULE in #7371

**Comment by kripken** [comment]
Oh good point, I forgot that... well, I guess we need a very different approach here then...

**Comment by sbc100** [comment]
This is all very confusing.   I would have thought that EXPORT_ALL should be enabled by default with MAIN_MODULE=1 and disabled with MAIN_MODULE=2.    Does that make sense?

So I think that #7371 maybe should have only effected the =1 case, and not the =2 case?

**Comment by kripken** [comment]
Yeah, the behavior of `=2` might need to be different here.

The Module object is used to interface between main and side modules, is perhaps the tricky thing here. So exporting onto Module, which is what `EXPORT_ALL` does, can assist dynamic linking. But fastcomp does this differently, it just adds them all on the env argument for imports to wasm, which is more direct, and more compact. Perhaps we should just do that for upstream too, and maybe that should only be done in `=1`. In `=2` the user would need to specify which JS library things to export, using `EXPORTED_RUNTIME_METHODS`. So this would generalize the 1/2 behavior of "do it all vs let the user specify which" from exporting compiled code to JS library code. What do you think?

**Comment by sbc100** [comment]
I agree that approach.

I'm still a little confused about the downside of `EXPORT_ALL` in the case of `MAIN_MODULE=1` because I don't understand what you mean by "fastcomp does this differently, it just adds them all on the env argument for imports to wasm, which is more direct".  Is this a potential missing optimization for the llvm backend path?  Or does it not make sense there?

It sounds like t here are two differnt levels of export here and that EXPORT_ALL is controlling both of them?

File: emcc.py [hunk 1]
```diff
@@ -1153,6 +1153,13 @@ def check(input_file):
       assert not shared.Settings.SIDE_MODULE
       if shared.Settings.MAIN_MODULE != 2:
         shared.Settings.INCLUDE_FULL_LIBRARY = 1
+      # fastcomp adds all the JS library to the asmLibraryArg object which is
+      # provided as "env" to wasm modules, which lets them access those methods.
+      # in the wasm backend we use a more straightforward method of flipping
+      # the EXPORT_ALL flag, which provides those methods on Module, which is
+      # used to look up imports as necessary.
+      if shared.Settings.WASM_BACKEND:
+        shared.Settings.EXPORT_ALL = 1
```
Comment (sbc100) [inline emcc.py]: I think we need to do this both for SIDE_MODULE=1 and MAIN_MODULE=1,  down below on ling 1170 would make sense to me.

Comment (sbc100) [reply emcc.py]: This still needs doing I think.

File: emcc.py [hunk 2]
```diff
@@ -1151,11 +1151,14 @@ def check(input_file):
...
       assert not shared.Settings.SIDE_MODULE
-      if shared.Settings.MAIN_MODULE != 2:
+      if shared.Settings.MAIN_MODULE == 1:
         shared.Settings.INCLUDE_FULL_LIBRARY = 1
+        shared.Settings.EXPORT_ALL = 1
     elif shared.Settings.SIDE_MODULE:
       assert not shared.Settings.MAIN_MODULE
       options.memory_init_file = False # memory init file is not supported with asm.js side modules, must be executable synchronously (for dlopen)
+      if not shared.Settings.WASM_BACKEND:
+        shared.Settings.EXPORT_ALL = 1
```
Comment (sbc100) [inline emcc.py]: I think you can just do this once down on line 1166, no?   Right below where we set `LINKABLE = 1`?

**Comment by kripken** [comment]
Oh, I was wrong in part of the analysis before. I saw that fastcomp adds all the JS library methods to `asmLibraryArg`, and that upstream does not, and I thought that explained the difference. So a possible solution was to just do what fastcomp does, as mentioned earlier. However, that doesn't really help - it's still increasing code size, but worse, it also means the dynamic loader must look for methods on `asmLibraryArg` too (when `resolveSymbol` is called for things like function pointers). So I think it's better to just use `EXPORT_ALL` for this.

But, as agreed earlier, let's only do it in `MAIN_MODULE=1`, so that in mode 2 users can control what is exported manually.

Also stop doing EXPORT_ALL in the test runner: do it in emcc.py when needed

File: src/support.js [hunk 3]
```diff
@@ -508,6 +508,11 @@ function loadWebAssemblyModule(binary, flags) {
           return obj[prop] = function() {
             if (!fp) {
               var f = resolveSymbol(name, 'function', legalized);
+#if ASSERTIONS
+              // this resolved symbol must exist, as we are about to try to
+              // add it to the table; error clearly here instead of there.
+              assert(f, 'could not resolve: ' + name);
```
Comment (sbc100) [inline src/support.js]: resolveSymbol already includes an assert which more detail than this.

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py]: Still needs doing.

File: src/support.js (hunk 3 above)
Comment (kripken) [reply src/support.js]: Thanks, I forgot to build with assertions when I tested this I think...

File: emcc.py (hunk 2 above)
Comment (kripken) [reply emcc.py]: Thanks, done.

File: emcc.py [hunk 4]
```diff
@@ -1154,16 +1154,18 @@ def check(input_file):
...
-      options.memory_init_file = False # memory init file is not supported with asm.js side modules, must be executable synchronously (for dlopen)
+      # memory init file is not supported with asm.js side modules, must be executable synchronously (for dlopen)
+      options.memory_init_file = False
 
     if shared.Settings.MAIN_MODULE or shared.Settings.SIDE_MODULE:
       assert shared.Settings.ASM_JS, 'module linking requires asm.js output (-s ASM_JS=1)'
-      if shared.Settings.MAIN_MODULE != 2 and shared.Settings.SIDE_MODULE != 2:
+      if shared.Settings.MAIN_MODULE == 1 or shared.Settings.SIDE_MODULE == 1:
         shared.Settings.LINKABLE = 1
+        shared.Settings.EXPORT_ALL = 1
```
Comment (sbc100) [inline emcc.py]: If WASM_BACKEND?

Comment (kripken) [reply emcc.py]: I think there is no harm to do it in fastcomp too, and it makes the two paths more similar. Also, I'm not sure that there isn't a fastcomp bug that isn't fixed by this, but I didn't investigate. But maybe I'm forgetting something?

**Pull request merged**